compiler/src/dmd/expressionsem.d

```diff
@@ -12796,6 +12796,59 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return;
         }
 
+        // Inline the expression, if possible.
+        BinExp be = cast(BinExp)e;
```

Comment:
Since `PowExp` is a subclass of `BinExp`, is there a reason to explicitly cast it to `be` instead of using `exp.e1` and `exp.e2` directly?

Reply:
I didn't use `exp.e1` and `exp.e2` because `optimize` converts `e2` to an `IntegerExp` (if possible) that we check for in the second condition in `be.e1.type.isScalar() && be.e2.isIntegerExp()`.

I didn't like the cast so I thought that instead of casting to `PowExp`, I cast to the lowest common denominator that has `e1` and `e2` (`BinExp`) but thinking about it now this is just bad analysis on my part.

Reply:
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.

Is it correct, that way?

Reply:
Oh, I just noticed that `e` is different from `exp` at that point because it's the return value of `exp.optimize`. In that case either a cast to BinExp or PowExp is fine.